lib/protocol/protocol.go

```diff
@@ -421,6 +420,37 @@ func (c *rawConnection) readMessage() (message, error) {
 	return msg, nil
 }
 
+func (c *rawConnection) readHeader() (Header, error) {
+	// First comes a 2 byte header length
+
+	if len(c.readerBuf) < 2 {
```

Comment:
also, this whole 
```
if c.buf < y {
  c.buf = make()
}
use(c.buf)
```

should be which deals allocations and potentially resizing.

use(c.getBuf(123))